sdk/python/kubeflow/training/api/training_client.py

```diff
@@ -81,6 +89,141 @@ def __init__(
             )
         self.job_kind = job_kind
 
+    def train(
+        self,
+        name: str = None,
+        namespace: str = None,
+        num_workers: int = 1,
+        num_procs_per_worker: int = 1,
+        storage_config: Dict[Literal["size", "storage_class"], str] = None,
```

Comment:
Should we initially just remove this parameter and rely on [default Storage Class](https://kubernetes.io/docs/concepts/storage/storage-classes/#default-storageclass) to provision PV @deepanker13 @johnugeorge ?
We might need to have discussions whether any Kubernetes-specific parameters are necessary for this API.

Reply:
The default storage class is "default". So user can skip it. However, the default storage class might not have ReadMany accessmode. In that case, they have to provide an appropriate storage class. I am assuming that admin will provide this detail. It is not ideal. But, I don't see a better way.

Reply:
@johnugeorge You can change the default storage class: https://kubernetes.io/docs/tasks/administer-cluster/change-default-storage-class/#changing-the-default-storageclass

Reply:
@andreyvelich Yes. we can change the defaults. However, most users like to have volumes as the default due to lesser cost. The default storage class in most of the providers does not have ReadWriteMany. Hence, I was thinking, storage config to be an advanced setting which is used only if user needs to override the default storage class(`default`).  So, it can be skipped in the general cases. What do you think?

Reply:
Actually, I was thinking more about it. Should the PVC creation be part of Training Operator controller rather than SDK?
In that case we can avoid giving unnecessary permission to Kubeflow User and make it more general for PyTorchJob APIs.

E.g. we need to introduce changes to API to be able to set the following in PyTorchJob:
```yaml
apiVersion: "kubeflow.org/v1"
kind: PyTorchJob
metadata:
  name: pytorchjob-yaml
  namespace: notebooks-test
spec:
  pytorchReplicaSpecs:
     .....
  storageSpec:
     storageClassName: my-sc
     resources:
        request:
           storage: 8Gi
```

So the parameter looks like this:
```go
type PyTorchJobSpec struct {
  storageSpec *corev1.PersistentVolumeClaimSpec `json:"storageSpec,omitempty"`
   ...
```

Any thoughts @kubeflow/wg-training-leads @kuizhiqing @deepanker13 ?

Reply:
cc @droctothorpe for feedback since your team worked on SDK changes for Kubeflow Training Operator and Katib

Reply:
I tend to favor optimizing for configurability, since it provides an escape hatch for corner cases that would otherwise require hot-fixes. There are countless storage classes and no way to enforce consistency across clusters and providers.

Simplicity can be preserved with sensible defaults. I imagine that would translate to no PVC in this particular context. `storage_config` can be typed as `Optional`.

Another variable to take into consideration is the mount point on the distributed pods. If you use EFS, for example, for its NFS capabilities, the performance is abysmal, so you likely want to restrict the mount to a specific subfolder.

[Here's](https://github.com/kubeflow/pipelines/blob/9a306129f8d33cdd0dc63dd10e87e51859b33eba/kubernetes_platform/python/kfp/kubernetes/volume.py#L26) KFP V2's implementation. As you can see, it has significantly more options.

How do we want to handle PVC garbage collection? Should there be dedicated PVC CRUD capabilities in the training SDK? Should the PyTorchJobSpec have a field that optionally tells the controller to delete the dynamically created PVC once the job is over?

Internally, we wrote a notebook and corresponding volume SDK (to support workspace and data volume CRUD). We really should contribute it.

Reply:
> How do we want to handle PVC garbage collection? Should there be dedicated PVC CRUD capabilities in the training SDK? Should the PyTorchJobSpec have a field that optionally tells the controller to delete the dynamically created PVC once the job is over?

I think, we should set PyTorchJob [OwnerReference ](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/) to make sure that Kubernetes cleanup appropriate resources when job is deleted.

@droctothorpe So what are thoughts on that PVC orchestration should be part of Training Operator controller reconcile loop ?

Reply:
The only thing about setting the OwnerReference is that you may want volumes to persist beyond the lifetime of the CR, so it might make sense to add a toggle for that similar to `CleanPodPolicy`.

Making the training controller create the PVC if it doesn't exist makes sense, IMO. It might also make sense to add some client-side PVC CRUD capabilities as well so that users can create, read, and delete PVCs out of band as well, since volume life-cycle management is unique.

Reply:
We discussed with Johnu how we should move forward here.
Initially, let's keep the PVC creation as part of SDK APIs.
Later we need to submit separate proposal to see how we can update the Job Spec API to support PVC creation inside the Training Operator controller.
That will allow to remove unnecessary permission from the user RBAC and unify the storage creation for all Training Operator Jobs (we can start with PyTorchJob).

Reply:
@johnugeorge @deepanker13 Please can you submit an issue to track this improvement ?

Reply:
#1971